pkg/controllers/jobset_controller.go

```diff
@@ -286,6 +246,197 @@ func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, js *
+			LastTransitionTime: metav1.Now(),
+			Reason:             "FailedJobs",
+			Message:            "jobset failed due to one or more failed jobs",
+		}); err != nil {
+			log.Error(err, "error updating status")
+			return err
+		}
+		return nil
+	}
+
+	// Handle different types of failure policy targets.
+	switch js.Spec.FailurePolicy.Target {
+	case jobset.TerminationPolicyTargetAny:
+		return r.failureTargetAny(ctx, js, ownedJobs)
+	case jobset.TerminationPolicyTargetAll:
+		return fmt.Errorf("unimplemented")
+	default:
+		return fmt.Errorf("invalid termination policy target %s", js.Spec.FailurePolicy.Target)
+	}
+}
+
+func (r *JobSetReconciler) failureTargetAny(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	// If any job from the target ReplicatedJob has failed, fail the JobSet.
+	// TODO: can we refactor using a different data structure for ownedJobs to avoid these nested for loops?
+	for _, target := range js.Spec.FailurePolicy.Jobs {
+		for _, failedJob := range ownedJobs.failed {
+			// Name of ReplicatedJob used to create this job is stored as a label on the job.
+			rjobName, ok := failedJob.Labels[jobset.ReplicatedJobLabel]
+			if !ok {
+				return fmt.Errorf("job is missing label %s", jobset.ReplicatedJobLabel)
+			}
+			// If we found a failed job that was created from the target replicated job,
+			// we trigger the failure policy and execute the restart policy.
+			if rjobName == target {
+				return r.executeRestartPolicy(ctx, js, ownedJobs)
+			}
+		}
+	}
+	return nil
+}
+
+func (r *JobSetReconciler) executeRestartPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	switch js.Spec.FailurePolicy.RestartPolicy {
+	case jobset.RestartPolicyRecreate:
+		return r.restartPolicyRecreate(ctx, js, ownedJobs)
+	case jobset.RestartPolicyNone:
+		return fmt.Errorf("unimplemented")
+	default:
+		return fmt.Errorf("invalid restart policy: %s", js.Spec.FailurePolicy.RestartPolicy)
+	}
+}
+
+func (r *JobSetReconciler) restartPolicyRecreate(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("jobset", klog.KObj(js))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// If JobSet status is already Restarting, we can exit early.
+	if isJobSetRestarting(js) {
+		log.V(2).Info("jobset is already restarting")
+		return nil
+	}
+
+	// First mark jobset as restarting to avoid multiple reconciliation
+	// calls attempting to delete the same jobs.
+	if err := r.updateStatus(ctx, js, metav1.Condition{
+		Type:               string(jobset.JobSetRestarting),
+		Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+		LastTransitionTime: metav1.Now(),
+		Reason:             "Restarting",
+		Message:            "jobset is currently restarting as part of recreation restart policy",
+	}); err != nil {
+		return fmt.Errorf("updating jobset status to restarting while executing recreation restart policy: %v", err)
+	}
+
```

Comment:
Imagine the controller is restarted right here after updating the status but before it gets to delete all jobs below, or in the middle of the loop issuing job deletions inside deleteAllJobs; in this case, the jobset will be stuck since at 316 above we will exit early.

I feel this approach of depending on the condition to determine what to do is brittle, also the semantics of "restarting" condition is not clear what it means from the user perspective (when do we exit it?)

I suggest we label the child jobs with the restart count to determine which jobs should be counted against the status of the jobset and which should not and be deleted. This approach will be more robust against controller restarts and race conditions.

The algorithm may look like this:

1. Child Jobs are created with label `jobset.sigs.k8s.io/restart-attampt` = `jobset.status.restarts`
2. On each reconcile, to determine success/failure, we only evaluate the status of jobs with `jobset.sigs.k8s.io/restart-attempt` == `jobset.status.restarts`
3. If there is failure, then increment `jobset.status.restarts`
4. Jobs where jobset.sigs.k8s.io/restart-attampt < jobset.status.restarts should be deleted.

Reply:
Good point! I wasn't thinking about the race conditions that could arise as a result of controller restarts mid-execution, though I can see how important that is especially in the k8s world where pods are somewhat ephemeral and may be restarted/moved to another node/etc. at any time.

I've implemented the suggested approach, please take a look and let me know if you have any additional thoughts.